### PR TITLE
feat: Support ESLint 7.x

### LINKED ISSUE #138: Support ESLint 7.x
ESLint v7.0.0 is [released](https://eslint.org/blog/2020/05/eslint-v7.0.0-released) 🎉

It would be awesome to have official ESLint 7 support. 👊
I'm happy to help where I can of course 🙂

### PR DESCRIPTION
**Pull request by MichaelDeBoey**
ESLint v7.0.0 is [released](https://eslint.org/blog/2020/05/eslint-v7.0.0-released) 🎉

(dev)Dependencies should be compatible with ESLint 7 too before we can merge this one:

- [x] [`@typescript-eslint/experimental-utils`](https://github.com/typescript-eslint/typescript-eslint/issues/1550)
  - [x]  [Release](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v3.0.0)
- [x] [`@typescript-eslint/eslint-plugin`](https://github.com/typescript-eslint/typescript-eslint/issues/1550)
  - [x]  [Release](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v3.0.0)
- [x] [`@typescript-eslint/parser`](https://github.com/typescript-eslint/typescript-eslint/issues/1550)
  - [x]  [Release](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v3.0.0)
- [x] [`eslint-plugin-import`](https://github.com/benmosher/eslint-plugin-import/issues/1772)
  - [x]  [PR](https://github.com/benmosher/eslint-plugin-import/pull/1715) + [commit](https://github.com/benmosher/eslint-plugin-import/commit/6a110dd16a7fd775f08601054bf14ffd503eea7b)
  - [x]  [Release](https://github.com/benmosher/eslint-plugin-import/blob/master/CHANGELOG.md#2211---2020-06-07)
- [x] [`eslint-plugin-jest`](https://github.com/jest-community/eslint-plugin-jest/issues/568)
  - [x]  [PR](https://github.com/jest-community/eslint-plugin-jest/pull/572)
  - [x]  Released
- [x] [`eslint-plugin-jest-formatting`](https://github.com/dangreenisrael/eslint-plugin-jest-formatting/issues/88)
  - [x]  [PR](https://github.com/dangreenisrael/eslint-plugin-jest-formatting/pull/89)
  - [x]  [Release](https://github.com/dangreenisrael/eslint-plugin-jest-formatting/releases/tag/v2.0.0)
- [ ] [`eslint-plugin-node`](https://github.com/mysticatea/eslint-plugin-node/issues/223)
  - [ ]  [PR](https://github.com/mysticatea/eslint-plugin-node/pull/224)
  - [ ]  Released
- [x] [`eslint-plugin-prettier`](https://github.com/prettier/eslint-plugin-prettier/issues/292)
  - [x]  [PR](https://github.com/prettier/eslint-plugin-prettier/issues/291)
  - [x]  Released
- [ ] [`eslint-plugin-promise`](https://github.com/xjamundx/eslint-plugin-promise/issues/189)
  - [ ]  [PR](https://github.com/xjamundx/eslint-plugin-promise/pull/190)
  - [ ]  Released
- [x] [`eslint-plugin-standard`](https://github.com/standard/eslint-plugin-standard/issues/35)
  - [x]  [PR](https://github.com/standard/eslint-plugin-standard/pull/36)
  - [x]  Released

---

Closes #138

### CODE REVIEW

**Comment by Belco90** [comment]
@MichaelDeBoey After rechecking this, I'm wondering if what you want to do is to force the plugin to work only for ESLint >= 7? If so, I would avoid this restriction and just make it compatible with ESLint 7 without losing compatibility with ESLint 6 at least. From your changes it looks it would be only compatible with ESLint 7.

**Comment by MichaelDeBoey** [comment]
@Belco90 `@typescript-eslint/experimental-utils` needs ESLint 7.x

**Comment by Belco90** [comment]
Sorry, let me reword this. There are two different uses of ESLint here:
- Plugin dev dependency itself. This is the version we install for developing the actual plugin. This one is fine to be upgraded to v7 [as you did in package.json](https://github.com/testing-library/eslint-plugin-testing-library/pull/139/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R50)
- Users dependency. This is the version of ESLint the users installing `eslint-plugin-testing-library` are using in their environments, and we don't have control over that. If we force the users to use ESLint v7 in order to make this plugin to work, all dependents projects using this plugin would be forced to migrate to ESLint v7 in order to keep using it. [You bumped this one too in package.json](https://github.com/testing-library/eslint-plugin-testing-library/pull/139/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R70)

So upgrading ESLint to v7 in our dev dependencies is fine, but it's not for peer dependency. ESLint is never a direct dependency of the plugin, so we should leave it compatible for ESLint >= 5 in peer dependency. This is why in the original issue I mentioned the fact we would need to update the CI to run all the tests against ESLint v5, v6 and v7 to make sure we offer proper compatibility for users using that range of versions in their projects.

You can check this in [eslint-plugin-react](https://github.com/yannickcr/eslint-plugin-react/blob/master/package.json). They don't even force ESLint to v7 in their dev dependencies but have a matrix instead with all the ESLint versions they offer compatibility in both dev and peer dependencies.

I don't know if I'm missing something else here or from your changes.

**Comment by MichaelDeBoey** [comment]
@Belco90 Like I said before: `@typescript-eslint/experimental-utils` (which is a dependency of this plugin) requires ESLint 7.x, so we need to force users of this plugin to use ESLint 7.x

Sorry if I wasn't clear enough that `@typescript-eslint/experimental-utils` is a dependency of this plugin 😕

**Comment by Belco90** [comment]
That's what I was missing. I though `@typescript-eslint/experimental-utils` was a dev dependency, but it's a regular dependency. We could add then ESLint v7 as direct dependency so it's bundled with the plugin and the user can use whatever ESLint version they want from 5 to 7.

But that will make the plugin really heavy, so it doesn't sound like a good option.

I don't know which other options we have (not upgrading `@typescript-eslint/experimental-utils` yet so we don't need ESLint v7 as dependency?) but definitely I do not want to force users to migrate to ESLint v7. We are not "adding support for ESLint 7.x" with the current PR as its title claims, but "migrating the whole plugin to ESLint 7.x" with its corresponding breaking change.

**Comment by Belco90** [comment]
Is `@typescript-eslint/experimental-utils` only compatible with ESLint v7 tho? I don't see that as breaking change, but just they added support for v7.

**Comment by Belco90** [comment]
[They support ESLint v5, v6 and v7](https://github.com/typescript-eslint/typescript-eslint#supported-eslint-version) so again unless I'm missing something, I don't think we need to restrict it to ESLint v7 here.

**Comment by MichaelDeBoey** [comment]
I'm sorry, I thought only ESLint 7.x was supported anymore.

I think forcing users to use ESLint 7.x as a breaking change is a good thing tbh.
We'll make the community use newer versions of software, which is a good thing imo.
Since this is a breaking change and it's the only breaking thing, I think this can be done.
Because technically we already support ESLint 7.x if we look at the current `peerDependencies`.

**Comment by Belco90** [comment]
Indeed, current peerDependencies are wrongly set so we need to fix that so it would be like `"eslint": "^5 || ^6 || ^7".

I don't think we should drop support for ESLint 5 and 6 right now. Packages just don't drop support for node v10 and v12 a month after node v14 is released. We should do the same with ESLint here, and it's what heavily used ESLint plugins are doing (like `typescript-eslint` or `eslint-plugin-react`).

Having said that, the changes for supporting ESLint v5, v6 and v7 aren't that complicated. Apart from the amazing work you did in this PR, we would need to update peer deps as mentioned before and add the ESLint dependencies matrix to CI. I can do the latter during this weekend, I have some spare time and it's something should have been included while ago to test support for v5 and v6. Probably updating the contributing guidelines is a good idea too.

At the moment we are not doing specific things for v7, so in the future we may need some extra checks in the code to do only few things depending on user ESLint version, but we are safe in that side for now.

**Comment by Belco90** [comment]
CI matrix added in #164. Could you take a look at it please?

**Comment by Belco90** [comment]
Merged. The ESLint matrix is available and ESLint versions have been fixed properly between v5 and v6. You need to update that here so it accepts ESLint versions between v5 and v7 and generate the matrix until v7. Let me know if I can help you with it.

**Comment by MichaelDeBoey** [comment]
@Belco90 I've rebased this branch onto `master/upstream` and deleted the breaking change notice, since you already updated that in your CI PR.

**Comment by Belco90** [comment]
Nice! The CI is passing: https://travis-ci.org/github/testing-library/eslint-plugin-testing-library/builds/698191193

Do we really need to wait for those dependencies from the description to be updated to v7 support?

**Comment by MichaelDeBoey** [comment]
@Belco90 Since they're not compatible with ESLint 7 yet, I would wait for them.

**Comment by Belco90** [comment]
I think we should leave peer dependencies with no ended range: https://twitter.com/SemanticRelease/status/1249795728253353986

**Review by nickmccurdy — DISMISSED** [review]

**Comment by chrisbobbe** [comment]
I see this update in the GitHub UI 🙂:

> nickmccurdy approved these changes on Jul 5

What's next for this PR; can it be merged, or maybe un-marked as a draft?

**Comment by nickmccurdy** [comment]
It seems to be blocked by https://github.com/mysticatea/eslint-plugin-node/issues/223 and https://github.com/xjamundx/eslint-plugin-promise/issues/189 updates

**Comment by Belco90** [comment]
Yep, although I've thought couple of times it would be fine to merge it. I think our plugin should work fine, it could imply just a bigger bundle size.

File: .travis.yml
Comment (Belco90) [inline .travis.yml:7]: Why is this back again to just testing ESLint v7.X? As previously discussed, this should maintain compatibility with ESLint >= 5 so we need v5, v6 and v7 in the matrix.

File: package.json
Comment (Belco90) [inline package.json:70]: Same as last comment: per previous discussion this should be any ESLint version greater or equal than 5
```suggestion
    "eslint": ">=5.0.0"
```

**Review by Belco90 — CHANGES_REQUESTED** [review]
Some regression changes around ESLint version declaration. I'm checking the failing test meanwhile.

File: .travis.yml
Comment (MichaelDeBoey) [reply .travis.yml:7]: I did this on purpose, since `@typescript-eslint/experimental-utils@4.x` requires `eslint@^7.5.0` in order to have the optional chaining support

File: package.json
Comment (MichaelDeBoey) [reply package.json:70]: I did this on purpose, since `@typescript-eslint/experimental-utils@4.x` requires `eslint@^7.5.0` in order to have the optional chaining support

Comment (Belco90) [reply package.json:70]: Not sure about this. This will force all users to migrate their ESLint to a version >= 7.5 and that's a huge limitation. As per private discussion, CRA is doing so, so probably we are fine?

**Comment by Belco90** [comment]
I can see several warnings after bumping TS to v4:

> ts-jest[versions] (WARN) Version 4.0.3 of typescript installed has not been tested with ts-jest. If you're experiencing issues, consider using a supported version (>=3.4.0 <4.0.0). Please do not report issues in ts-jest if you are using unsupported versions.

It doesn't solve the issues in the tests tho, but I'm not sure if we should bump TS then.

File: package.json
Comment (MichaelDeBoey) [reply package.json:70]: CRA is doing so indeed and it's needed for optional chaining support too.

**Comment by MichaelDeBoey** [comment]
Since all tests are green for ESLint 7.x, I think we can merge this one without https://github.com/mysticatea/eslint-plugin-node/pull/224 & https://github.com/xjamundx/eslint-plugin-promise/pull/190 being merged 🙂

**Review by Belco90 — CHANGES_REQUESTED** [review]
Happy to merge this, but since it is a breaking change I'm gonna request to point this PR to v4 rather than master. Can you do that please? So we use that new major to introduce this breaking change. I think you shouldn't have any issue pointing to that branch. Let me know if you need help if you find any problem tho.

Also, can you please update Installation section from README to specify ESLint v7.5 is required now?

Thanks.

**Comment by MichaelDeBoey** [comment]
@Belco90 I'll revert all `@typescript-eslint/experimental-utils` changes so we can merge this one and unblock CRA.
I'll create a new PR with the changes and point it to `v4` branch once this one's merged 🙂

**Comment by Belco90** [comment]
Brilliant. Thanks for your hard work @MichaelDeBoey

**Comment by MichaelDeBoey** [comment]
@Belco90 All tests are green 🙂

File: package.json
Comment (Belco90) [inline package.json:70]: This should be updated to greater or equal than 5.
```suggestion
    "eslint": ">=5.0.0"
```

**Review by Belco90 — CHANGES_REQUESTED** [review]
Awesome. Just a small tweak to peerDependencies and should be ready to go!

**Review by Belco90 — APPROVED** [review]

**Comment by Belco90** [comment]
@all-contributors please add @MichaelDeBoey for code and platform

**Pull request merged**
